File: sdks/cpp/src/session.cpp
#include "sibna/session.hpp"
#include "sibna/context.hpp"
#include "sibna/crypto.hpp"
#include "sibna/x3dh.hpp"

namespace sibna {

// ── Session ──────────────────────────────────────────────────────────────────

Session::Session(bytes peer_id, void* native_handle)
    : session_key_(Utils::random_bytes<KEY_LENGTH>())
    , peer_id_(std::move(peer_id))
    , native_handle_(native_handle)
{}

Session::~Session() {
    Utils::secure_clear(peer_id_);
    Utils::secure_clear(session_key_);
    disposed_ = true;
}

Session::Session(Session&& other) noexcept
    : session_key_(std::move(other.session_key_))
    , ratchet_(std::move(other.ratchet_))
    , ratchet_initialized_(other.ratchet_initialized_)
    , peer_id_(std::move(other.peer_id_))
    , native_handle_(other.native_handle_)
    , disposed_(other.disposed_)
    , messages_sent_(other.messages_sent_)
    , messages_received_(other.messages_received_)
    , established_at_(other.established_at_)
{
    other.native_handle_ = nullptr;
    other.disposed_ = true;
}

Session& Session::operator=(Session&& other) noexcept {
    if (this != &other) {
        Utils::secure_clear(peer_id_);
        Utils::secure_clear(session_key_);
        peer_id_ = std::move(other.peer_id_);
        native_handle_ = other.native_handle_;
        disposed_ = other.disposed_;
        messages_sent_ = other.messages_sent_;
        messages_received_ = other.messages_received_;
        established_at_ = other.established_at_;
        session_key_ = std::move(other.session_key_);
        ratchet_ = std::move(other.ratchet_);
        ratchet_initialized_ = other.ratchet_initialized_;
        other.native_handle_ = nullptr;
        other.disposed_ = true;
    }
    return *this;
}

Result<void> Session::perform_handshake(const PreKeyBundle& peer_bundle, bool /* initiator */) {
    ensure_not_disposed();

    // Validate the peer bundle
    if (peer_bundle.is_expired()) {
        return Result<void>(ResultCode::INVALID_ARGUMENT, "Peer bundle is expired");
    }

    // Verify the SPK signature before any DH computation
    auto sig_result = peer_bundle.verify_signature(peer_bundle.identity_key());
    if (sig_result.is_err() || !sig_result.value()) {
        return Result<void>(ResultCode::AUTHENTICATION_FAILED, "Bundle signature verification failed");
    }

    // In a full implementation, the caller would:
    // 1. Generate ephemeral X25519 keypair
    // 2. Call x3dh_initiator/x3dh_responder to get shared_secret
    // 3. Derive session keys via x3dh_derive_session_keys
    // 4. Initialize DoubleRatchet from shared secret

    established_at_ = std::chrono::system_clock::now();

    return Result<void>();
}

Result<bytes> Session::encrypt(const bytes& plaintext, const bytes& associated_data) {
    ensure_not_disposed();

    if (plaintext.empty()) {
        return Result<bytes>(ResultCode::INVALID_ARGUMENT, "Plaintext cannot be empty");
    }

    // Use Double Ratchet if initialized
    if (ratchet_initialized_) {
        auto result = ratchet_.ratchet_encrypt(plaintext, associated_data);
        if (result.is_ok()) {
            messages_sent_++;
        }
        return result;
    }

    // Fallback: direct encryption with session key
    auto encrypt_result = Crypto::encrypt(session_key_, plaintext, associated_data);
    if (encrypt_result.is_ok()) {
        messages_sent_++;
    }
    return encrypt_result;
}

Result<bytes> Session::decrypt(const bytes& ciphertext, const bytes& associated_data) {
    ensure_not_disposed();

    if (ciphertext.size() < NONCE_LENGTH + TAG_LENGTH + 1) {
        return Result<bytes>(ResultCode::INVALID_CIPHERTEXT, "Ciphertext too short");
    }

    // Use Double Ratchet if initialized
    if (ratchet_initialized_) {
        auto result = ratchet_.ratchet_decrypt(ciphertext, associated_data);
        if (result.is_ok()) {
            messages_received_++;
        }
        return result;
    }

    // Fallback: direct decryption with session key
    auto decrypt_result = Crypto::decrypt(session_key_, ciphertext, associated_data);
    if (decrypt_result.is_ok()) {
        messages_received_++;
    }
    return decrypt_result;
}

size_t Session::current_message_number() const {
    return messages_sent_;
}

bool Session::is_established() const {
    return established_at_.has_value();
}

std::optional<std::chrono::seconds> Session::age() const {
    if (!established_at_) {
        return std::nullopt;
    }
    auto now = std::chrono::system_clock::now();
    return std::chrono::duration_cast<std::chrono::seconds>(now - *established_at_);
}

SessionInfo Session::get_stats() const {
    SessionInfo info;
    info.peer_id = peer_id_;
    info.messages_sent = messages_sent_;
    info.messages_received = messages_received_;
    info.established_at = established_at_;
    info.is_established = established_at_.has_value();
    return info;
}

void Session::ensure_not_disposed() const {
    if (disposed_) {
        throw SibnaError(ResultCode::INVALID_STATE, "Session has been disposed");
    }
}

} // namespace sibna

File: sdks/go/session.go
package sibna

import (
	"fmt"
	"time"
)

// ── SessionConfig ──────────────────────────────────────────────────────────

type SessionConfig struct {
	MaxSkippedMessages int
	MaxChainMessages   int
}

func DefaultSessionConfig() *SessionConfig {
	return &SessionConfig{
		MaxSkippedMessages: MaxSkippedMessages,
		MaxChainMessages:   MaxChainMessages,
	}
}

// ── Session ────────────────────────────────────────────────────────────────

type Session struct {
	Config       *SessionConfig
	Ratchet      *DoubleRatchet
	SessionID    string
	EstablishedAt float64
	PeerID       []byte
}

func NewSession(config *SessionConfig) *Session {
	if config == nil {
		config = DefaultSessionConfig()
	}
	return &Session{
		Config:    config,
		SessionID: generateSessionID(),
	}
}

func (s *Session) IsEstablished() bool {
	return s.Ratchet != nil
}

func (s *Session) MessagesSent() int {
	if s.Ratchet != nil {
		return s.Ratchet.MessagesSent
	}
	return 0
}

func (s *Session) MessagesReceived() int {
	if s.Ratchet != nil {
		return s.Ratchet.MessagesReceived
	}
	return 0
}

// ── Initiator ──────────────────────────────────────────────────────────────

func (s *Session) InitiateAsInitiator(
	identity, ephemeral *X25519KeyPair,
	peerBundle *PreKeyBundle,
	ourDeviceID, peerDeviceID []byte,
) ([]byte, error) {
	result, err := X3DHInitiator(
		identity,
		ephemeral,
		peerBundle,
		ourDeviceID,
		peerDeviceID,
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("x3dh initiator: %w", err)
	}

	s.Ratchet, err = DoubleRatchetFromSharedSecret(
		result.SharedSecret,
		ephemeral.PrivateKeyBytes(),
		peerBundle.SignedPreKey,
		true,
	)
	if err != nil {
		return nil, fmt.Errorf("init ratchet: %w", err)
	}
	s.EstablishedAt = float64(time.Now().Unix())
	s.PeerID = append([]byte{}, peerBundle.IdentityKey...)

	return ephemeral.PublicKeyBytes(), nil
}

// ── Responder ──────────────────────────────────────────────────────────────

func (s *Session) InitiateAsResponder(
	identity, signedPrekey *X25519KeyPair,
	onetimePrekey *X25519KeyPair,
	peerIdentity, peerEphemeral []byte,
	ourDeviceID, peerDeviceID []byte,
) error {
	result, err := X3DHResponder(
		identity,
		signedPrekey,
		onetimePrekey,
		peerIdentity,
		peerEphemeral,
		ourDeviceID,
		peerDeviceID,
		nil,
	)
	if err != nil {
		return fmt.Errorf("x3dh responder: %w", err)
	}

	s.Ratchet, err = DoubleRatchetFromSharedSecret(
		result.SharedSecret,
		signedPrekey.PrivateKeyBytes(),
		peerEphemeral,
		false,
	)
	if err != nil {
		return fmt.Errorf("init ratchet: %w", err)
	}
	s.EstablishedAt = float64(time.Now().Unix())
	s.PeerID = append([]byte{}, peerIdentity...)
	return nil
}

// ── FromSharedSecret (testing) ─────────────────────────────────────────────

func SessionFromSharedSecret(
	sharedSecret, localDHPrivate, remoteDHPublic []byte,
	roleIsInitiator bool,
) (*Session, error) {
	ratchet, err := DoubleRatchetFromSharedSecret(
		sharedSecret,
		localDHPrivate,
		remoteDHPublic,
		roleIsInitiator,
	)
	if err != nil {
		return nil, err
	}
	return &Session{
		Config:        DefaultSessionConfig(),
		Ratchet:       ratchet,
		SessionID:     generateSessionID(),
		EstablishedAt: float64(time.Now().Unix()),
	}, nil
}

// ── Encrypt / Decrypt ─────────────────────────────────────────────────────

func (s *Session) Encrypt(plaintext, associatedData []byte) ([]byte, error) {
	if s.Ratchet == nil {
		return nil, fmt.Errorf("session not established")
	}
	return s.Ratchet.RatchetEncrypt(plaintext, associatedData)
}

func (s *Session) Decrypt(ciphertext, associatedData []byte) ([]byte, error) {
	if s.Ratchet == nil {
		return nil, fmt.Errorf("session not established")
	}
	return s.Ratchet.RatchetDecrypt(ciphertext, associatedData)
}

// ── Helpers ────────────────────────────────────────────────────────────────

func generateSessionID() string {
	b, _ := RandomBytes(16)
	return fmt.Sprintf("%x", b)
}
